README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Semi-Distance Correlation and MV Index: Measure Dependence Between Categorical and Continuous Variables

<!-- badges: start -->

[![R-CMD-check](https://github.com/wzhong41/semidist/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/wzhong41/semidist/actions/workflows/R-CMD-check.yaml)
<!-- badges: end --> The goal of package `semidist` is to provide an
easy way to implement the semi-distance methods (Zhong et al., 2023) and
MV index methods (Cui, Li and Zhong, 2015; Cui and Zhong, 2019).

## Installation

To install `semidist`,

``` r
install.packages("semidist")
```

## Example

Here is a simple example showing how to use `semidist` to measure the
dependence between a categorical variable and a multivariate continuous
variable, and apply the measure on testing the independence and conduct
groupwise feature screening.

``` r
library(semidist)
X <- mtcars[, c("mpg", "disp", "drat", "wt")]
y <- factor(mtcars[, "am"])

sdcov(X, y)
#> [1] 31.78288
sdcor(X, y)
#> [1] 0.3489821

sd_test(X, y)
#> 
#>  Semi-Distance Independence Test (Permutation Test with K = 10000)
#> 
#> Data: X and y,   Sample size = 32
#> Test statistic = 940.344,    p-value = 0.0005999
#> Alternative hypothesis: Two random variables are not independent

sd_sis(X, y, d = 2)
#> $group_info
#> $group_info$`Grp mpg`
#> [1] "mpg"
#> 
#> $group_info$`Grp disp`
#> [1] "disp"
#> 
#> $group_info$`Grp drat`
#> [1] "drat"
#> 
#> $group_info$`Grp wt`
#> [1] "wt"
#> 
#> 
#> $measurement
#>   Grp mpg  Grp disp  Grp drat    Grp wt 
#> 0.3447938 0.3488447 0.5054821 0.5358834 
#> 
#> $selected
#> [1] "wt"   "drat"
#> 
#> $ordering
#> [1] "Grp wt"   "Grp drat" "Grp disp" "Grp mpg"

# Suppose we have prior information for the group structure as
# ("mpg", "drat"), ("disp", "hp") and ("wt", "qsec")
group_info <- list(
  mpg_drat = c("mpg", "drat"),
  disp_wt = c("disp", "wt")
)
sd_sis(X, y, group_info, d = 2)
#> $group_info
#> $group_info$mpg_drat
#> [1] "mpg"  "drat"
#> 
#> $group_info$disp_wt
#> [1] "disp" "wt"  
#> 
#> 
#> $measurement
#>  mpg_drat   disp_wt 
#> 0.3518051 0.3488598 
#> 
#> $selected
#> [1] "mpg"  "drat"
#> 
#> $ordering
#> [1] "mpg_drat" "disp_wt"
```

## References

1.  Wei Zhong, Zhuoxi Li, Wenwen Guo and Hengjian Cui. (2023)
    “Semi-Distance Correlation and Its Applications.” *Journal of the
    American Statistical Association.*
2.  Hengjian Cui and Wei Zhong (2019). “A Distribution-Free Test of
    Independence Based on Mean Variance Index.” *Computational
    Statistics & Data Analysis*, 139, 117-133.
3.  Hengjian Cui, Runze Li and Wei Zhong (2015). “Model-Free Feature
    Screening for Ultrahigh Dimensional Discriminant Analysis.” *Journal
    of the American Statistical Association*, 110, 630-641.


NEWS.md
# semidist 0.1.0

* Initial CRAN submission.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(semidist)

test_check("semidist")


tests/testthat/test-utils.R
test_that("tr_estimate() works", {
  X <- matrix(rnorm(100), 20, 5)
  expect_equal(tr_estimate_R_impl(X), tr_estimate(X))
})


R/mv.R
#' Mean Variance (MV) statistics
#' @description Compute the statistics of mean variance (MV) index, which can
#'   measure the dependence between a univariate continuous variable and a
#'   categorical variable. See Cui, Li and Zhong (2015); Cui and Zhong (2019) 
#'   for details.
#'
#' @param x Data of univariate continuous variables, which should be a vector of
#'   length \eqn{n}.
#' @param y Data of categorical variables, which should be a factor of length
#'   \eqn{n}.
#' @param return_mat A boolean. If `FALSE` (the default), only the calculated
#'   statistic is returned. If `TRUE`, also return the matrix of the indicator
#'   for x <= x_i, which is useful for the permutation test.
#'
#' @return The value of the corresponding sample statistic.
#'
#'   If the argument `return_mat` of `mv()` is set as `TRUE`, a list with
#'   elements
#'   * `mv`: the MV index statistic;
#'   * `mat_x`: the matrices of the distances of the indicator for x <= x_i;
#'
#'   will be returned.
#' 
#' @seealso 
#' * [mv_test()] for implementing independence test via MV index;
#' * [mv_sis()] for implementing feature screening via MV index.
#' 
#' @examples
#' x <- mtcars[, "mpg"]
#' y <- factor(mtcars[, "am"])
#' print(mv(x, y))
#'
#' # Man-made independent data -------------------------------------------------
#' n <- 30; R <- 5; prob <- rep(1/R, R)
#' x <- rnorm(n)
#' y <- factor(sample(1:R, size = n, replace = TRUE, prob = prob), levels = 1:R)
#' print(mv(x, y))
#'
#' # Man-made functionally dependent data --------------------------------------
#' n <- 30; R <- 3
#' x <- rep(0, n)
#' x[1:10] <- 0.3; x[11:20] <- 0.2; x[21:30] <- -0.1
#' y <- factor(rep(1:3, each = 10))
#' print(mv(x, y))
#' 
#' @export
mv <- function(x, y, return_mat = FALSE) {
  if (is.data.frame(x)) {
    x <- as.matrix(x)
  }
  
  if (is.matrix(x) && ncol(x) > 1) {
    stop("`x` should be a numeric vector")
  } else {
    x <- drop(x)
  }
  
  n <- length(x)
  R <- nlevels(y)
  Y <- switch_cat_repr(y)
  n_all <- colSums(Y)
  
  # Column i stands for the indicator for x <= x[i]
  mat_x <- vapply(
    x, 
    function(x_i) x <= x_i, 
    rep(1, n)
  )
  
  # Marginal CDF of x (evaluated at each x_i) 
  F_0 <- colMeans(mat_x)
  # F_0 <- rank(x, ties.method = "max") / n # Equivalent!
  
  # Conditional CDF of x given y = r (evaluated at each x_i) 
  # Column r stands for the F_r
  F_all <- crossprod(mat_x, Y) %*% diag(1 / n_all)
  
  mv <- sum((F_all - F_0)^2 %*% diag(n_all/n)) / n
  
  if (return_mat) {
    return(
      list(
        mv = mv,
        mat_x = mat_x
      )
    )
  } else {
    return(mv)
  }
}


# Critical value via asymptotic distribution of MV statistic
mv_crit_asym <- function(R, N = 500, realizations = 10000) {
  x <- numeric(realizations)
  hm <- sum(1 / (1:N)^2) # compute the general harmonic number H_{500, 2}
  for (i in 1:realizations){
    xi <- 0
    for (j in 1:N){
      xi <- xi + rchisq(1, R-1) / j^2
    }
      x[i] <- xi / pi^2 + (R-1) * (1/6 - hm / pi^2)  # (R-1)*(1/6-hm/pi^2) is the correction error
  }
  quantile(x, probs = c(0.9, 0.95, 0.99))
}


#' MV independence test
#'
#' @description Implement the MV independence test via permutation test, or via 
#' the asymptotic approximation
#'
#' @param x Data of univariate continuous variables, which should be a vector of
#'   length \eqn{n}.
#' @param y Data of categorical variables, which should be a factor of length
#'   \eqn{n}.
#' @param test_type Type of the test:
#'   * `"perm"` (the default): Implement the test via permutation test;
#'   * `"asym"`: Implement the test via the asymptotic approximation.
#'   
#'   See the Reference for details.
#' @param num_perm The number of replications in permutation test.
#'
#' @returns A list with class `"indtest"` containing the following components
#' * `method`: name of the test;
#' * `name_data`: names of the `x` and `y`;
#' * `n`: sample size of the data;
#' * `num_perm`: number of replications in permutation test;
#' * `stat`: test statistic;
#' * `pvalue`: computed p-value. (Notice: asymptotic test cannot return a p-value, but only the critical values `crit_vals` for 90%, 95% and 99% confidence levels.)
#'
#' @examples
#' x <- mtcars[, "mpg"]
#' y <- factor(mtcars[, "am"])
#' test <- mv_test(x, y)
#' print(test)
#' test_asym <- mv_test(x, y, test_type = "asym")
#' print(test_asym)
#'
#' # Man-made independent data -------------------------------------------------
#' n <- 30; R <- 5; prob <- rep(1/R, R)
#' x <- rnorm(n)
#' y <- factor(sample(1:R, size = n, replace = TRUE, prob = prob), levels = 1:R)
#' test <- mv_test(x, y)
#' print(test)
#' test_asym <- mv_test(x, y, test_type = "asym")
#' print(test_asym)
#'
#' # Man-made functionally dependent data --------------------------------------
#' n <- 30; R <- 3
#' x <- rep(0, n)
#' x[1:10] <- 0.3; x[11:20] <- 0.2; x[21:30] <- -0.1
#' y <- factor(rep(1:3, each = 10))
#' test <- mv_test(x, y)
#' print(test)
#' test_asym <- mv_test(x, y, test_type = "asym")
#' print(test_asym)
#'
#' @export
mv_test <- function(x, y, test_type = "perm", num_perm = 10000) {
  mv_obj <- mv(x, y, return_mat = TRUE)
  
  n <- length(y)
  Y <- switch_cat_repr(y)
  R <- ncol(Y)
  n_all <- colSums(Y)
  name_data <- paste(
    deparse(substitute(x)), "and", deparse(substitute(y))
  )
  
  mv_n <- mv_obj$mv
  mat_x <- mv_obj$mat_x
  F_0 <- colMeans(mat_x)
  
  if (test_type == "perm") {
    name_method <- paste0(
      "MV Independence Test (Permutation Test with K = ",
      num_perm, ")"
    )
    num_rej <- 0
    for (perm in 1:num_perm) {
      F_all <- crossprod(mat_x, Y[sample(1:n), ]) %*% diag(1 / n_all)
      
      mv_perm <- sum((F_all - F_0)^2 %*% diag(n_all/n)) / n
      
      if (mv_perm >= mv_n) {
        num_rej <- num_rej + 1
      }
    }
    pvalue <- (num_rej + 1) / (num_perm + 1)
  } else if (test_type == "asym") {
    name_method <- paste0(
      "MV Independence Test (Asymptotic Test)"
    )
    crit_mat <- matrix( # Taken from the Table 2 in Cui and Zhong (2019)
      c(
        0.3469, 0.6086, 0.8384, 1.0602, 1.2698, 1.4927, 1.6780, 1.8963, 2.1019,
        0.4665, 0.7365, 0.9938, 1.2399, 1.4440, 1.6986, 1.8901, 2.1096, 2.3362,
        0.7120, 1.0673, 1.3393, 1.6026, 1.8650, 2.1241, 2.3645, 2.6180, 2.7943
      ),
      nrow = 9, ncol = 3,
      dimnames = list(2:10, c("90%", "95%", "99%"))
    )
    if (R %in% 2:10) {
      crit_vals <- crit_mat[as.character(R), ]
    } else {
      crit_vals <- round(mv_crit_asym(R), 4)
    }
  }
  
  
  indtest <- list(
    method = name_method,
    name_data = name_data,
    n = n,
    stat = n * mv_n
  )
  if (test_type == "perm") {
    indtest$num_perm <- num_perm
    indtest$pvalue <- pvalue
  } else if (test_type == "asym") {
    indtest$crit_vals <- crit_vals
  }
  class(indtest) <- "indtest"
  
  indtest
}

#' Feature screening via MV Index
#'
#' @description Implement the feature screening for the classification problem
#'   via MV index.
#'
#' @param X Data of multivariate covariates, which should be an
#'   \eqn{n}-by-\eqn{p} matrix.
#' @param y Data of categorical response, which should be a factor of length
#'   \eqn{n}.
#' @param d An integer specifying how many features should be kept after
#'   screening. Defaults to `NULL`. If `NULL`, then it will be set as \eqn{[n /
#'   log(n)]}, where \eqn{[x]} denotes the integer part of x.
#'
#' @param parallel A boolean indicating whether to calculate parallelly via
#'   `furrr::future_map`. Defaults to `FALSE`.
#'
#' @returns A list of the objects about the implemented feature screening:
#' * `measurement`: sample MV index calculated for each single covariate;
#' * `selected`: indicies or names (if avaiable as colnames of `X`) of
#'   covariates that are selected after feature screening;
#' * `ordering`: order of the calculated measurements of each single covariate.
#'   The first one is the largest, and the last is the smallest.
#'
#' @examples
#' X <- mtcars[, c("mpg", "disp", "hp", "drat", "wt", "qsec")]
#' y <- factor(mtcars[, "am"])
#'
#' mv_sis(X, y, d = 4)
#'
#' @export
mv_sis <- function(X, y, d = NULL, parallel = FALSE) {
  p <- ncol(X)
  n <- nrow(X)
  
  if (is.null(d)) {
    d <- floor(n / log(n))
  }
  
  if (is.null(colnames(X))) {
    names_X <- 1:p
    colnames(X) <- names_X
  } else {
    names_X <- colnames(X)
  }
  
  omega <- furrr::future_map_dbl(names_X, ~ mv(X[, .x], y))
  ordering <- order(omega, decreasing = TRUE)
  ordering_names <- names_X[ordering]
  D_hat <- ordering_names[seq(d)]
  
  list(
    measurement = omega,
    selected = D_hat,
    ordering = ordering_names
  )
}


src/utils.cpp
#include <RcppArmadillo.h>
// [[Rcpp::depends(RcppArmadillo)]]

using namespace Rcpp;

//' Estimate the trace of the covariance matrix and its square
//' 
//' @description 
//' For a design matrix \eqn{\mathbf{X}}, estimate the trace of its covariance matrix \eqn{\Sigma = \mathrm{cov}(\mathbf{X})},
//' and the square of covariance matrix \eqn{\Sigma^2}.
//'
//' @param X The design matrix.
//' @return A list with elements:
//' * `tr_S`: estimate for trace of \eqn{\Sigma};
//' * `tr_S2`: estimate for trace of \eqn{\Sigma^2}.
// [[Rcpp::export]]
List tr_estimate(const arma::mat& X) {
  int n = X.n_rows;
  
  arma::mat XXt = X * X.t();
  double Y1 = arma::accu(arma::diagvec(XXt)) / n;
  double Y3 = (arma::accu(XXt) - Y1 * n) / (n * (n - 1));
  double Y2 = (arma::accu(arma::pow(XXt, 2)) - arma::accu(arma::pow(arma::diagvec(XXt), 2))) / (n * (n - 1));
  
  arma::mat XXt2 = XXt * XXt;
  arma::vec diag_XXt = arma::vectorise(arma::repelem(arma::diagvec(XXt), n - 1, 1));
  arma::vec offdiag_XXt = arma::vectorise(XXt(arma::find(arma::eye<arma::umat>(n, n) == 0)));
  double Y4 = (arma::accu(XXt2) - arma::accu(arma::diagvec(XXt2)) - 2 * arma::accu(diag_XXt % offdiag_XXt)) /
    (n * (n - 1) * (n - 2));
  double Y5 = ((n * (n - 1) * Y3) * (n * (n - 1) * Y3) - 2 * n * (n - 1) * Y2 - 4 * n * (n - 1) * (n - 2) * Y4) /
    (n * (n - 1) * (n - 2) * (n - 3));
  
  double Tn1 = Y1 - Y3;
  double Tn2 = Y2 - 2 * Y4 + Y5;
  
  return List::create(
    Named("tr_S") = Tn1,
    Named("tr_S2") = Tn2
  );
}

R/semidist.R
#' Semi-distance covariance and correlation statistics
#'
#' @description Compute the statistics (or sample estimates) of semi-distance
#'   covariance and correlation. The semi-distance correlation is a standardized
#'   version of semi-distance covariance, and it can measure the dependence
#'   between a *multivariate* continuous variable and a categorical variable.
#'   See Details for the definition of semi-distance covariance and
#'   semi-distance correlation.
#'
#' @param X Data of multivariate continuous variables, which should be an
#'   \eqn{n}-by-\eqn{p} matrix, or, a vector of length \eqn{n} (for univariate
#'   variable).
#' @param y Data of categorical variables, which should be a factor of length
#'   \eqn{n}.
#' @param type Type of statistic: `"V"` (the default) or `"U"`. See Details.
#' @param return_mat A boolean. If `FALSE` (the default), only the calculated
#'   statistic is returned. If `TRUE`, also return the matrix of the distances
#'   of X and the divergences of y, which is useful for the permutation test.
#'
#' @details For \eqn{\bm{X} \in \mathbb{R}^{p}} and \eqn{Y \in \{1, 2, \cdots,
#'   R\}}, the (population-level) semi-distance covariance is defined as
#'   \deqn{\mathrm{SDcov}(\bm{X}, Y) =
#'   \mathrm{E}\left[\|\bm{X}-\widetilde{\bm{X}}\|\left(1-\sum_{r=1}^R
#'   I(Y=r,\widetilde{Y}=r)/p_r\right)\right],} where \eqn{p_r = P(Y = r)} and
#'   \eqn{(\widetilde{\bm{X}}, \widetilde{Y})} is an iid copy of \eqn{(\bm{X},
#'   Y)}.
#'   The (population-level) semi-distance correlation is defined as
#'   \deqn{\mathrm{SDcor}(\bm{X}, Y) = \dfrac{\mathrm{SDcov}(\bm{X},
#'   Y)}{\mathrm{dvar}(\bm{X})\sqrt{R-1}},} where \eqn{\mathrm{dvar}(\bm{X})} is
#'   the distance variance (Szekely, Rizzo, and Bakirov 2007) of \eqn{\bm{X}}.
#'
#'   With \eqn{n} observations \eqn{\{(\bm{X}_i, Y_i)\}_{i=1}^{n}}, `sdcov()`
#'   and `sdcor()` can compute the sample estimates for the semi-distance
#'   covariance and correlation.
#'
#'   If `type = "V"`, the semi-distance covariance statistic is computed as a
#'   V-statistic, which takes a very similar form as the energy-based statistic
#'   with double centering, and is always non-negative. Specifically,
#'   \deqn{\text{SDcov}_n(\bm{X}, y) = \frac{1}{n^2} \sum_{k=1}^{n}
#'   \sum_{l=1}^{n} A_{kl} B_{kl},}
#'   where \deqn{A_{kl} = a_{kl} - \bar{a}_{k.} - \bar{a}_{.l} + \bar{a}_{..}}
#'   is the double centering (Szekely, Rizzo, and Bakirov 2007) of
#'   \eqn{a_{kl} = \| \bm{X}_k - \bm{X}_l \|,} and \deqn{B_{kl} =
#'   1 - \sum_{r=1}^{R} I(Y_k = r) I(Y_l = r) / \hat{p}_r} with \eqn{\hat{p}_r =
#'   n_r / n = n^{-1}\sum_{i=1}^{n} I(Y_i = r)}.
#'   The semi-distance correlation statistic is \deqn{\text{SDcor}_n(\bm{X}, y)
#'   = \dfrac{\text{SDcov}_n(\bm{X}, y)}{\text{dvar}_n(\bm{X})\sqrt{R - 1}},}
#'   where \eqn{\text{dvar}_n(\bm{X})} is the V-statistic of distance variance
#'   of \eqn{\bm{X}}.
#'
#'   If `type = "U"`, then the semi-distance covariance statistic is computed as
#'   an ``estimated U-statistic'', which is utilized in the independence test
#'   statistic and is not necessarily non-negative. Specifically,
#'   \deqn{\widetilde{\text{SDcov}}_n(\bm{X}, y) = \frac{1}{n(n-1)}
#'   \sum_{i \ne j} \| \bm{X}_i - \bm{X}_j \| \left(1 - \sum_{r=1}^{R}
#'   I(Y_i = r) I(Y_j = r) / \tilde{p}_r\right),}
#'   where \eqn{\tilde{p}_r = (n_r-1) / (n-1) = (n-1)^{-1}(\sum_{i=1}^{n} I(Y_i
#'   = r) - 1)}. Note that the test statistic of the semi-distance independence
#'   test is \deqn{T_n = n \cdot \widetilde{\text{SDcov}}_n(\bm{X}, y).}
#'
#' @returns The value of the corresponding sample statistic.
#'
#'   If the argument `return_mat` of `sdcov()` is set as `TRUE`, a list with
#'   elements
#'   * `sdcov`: the semi-distance covariance statistic;
#'   * `mat_x, mat_y`: the matrices of the distances of X and the divergences
#'   of y, respectively;
#'
#'   will be returned.
#'
#' @seealso
#'   * [sd_test()] for implementing independence test via semi-distance
#' covariance;
#'   * [sd_sis()] for implementing groupwise feature screening via
#' semi-distance correlation.
#'
#' @examples
#' X <- mtcars[, c("mpg", "disp", "drat", "wt")]
#' y <- factor(mtcars[, "am"])
#' print(sdcov(X, y))
#' print(sdcor(X, y))
#'
#' # Man-made independent data -------------------------------------------------
#' n <- 30; R <- 5; p <- 3; prob <- rep(1/R, R)
#' X <- matrix(rnorm(n*p), n, p)
#' y <- factor(sample(1:R, size = n, replace = TRUE, prob = prob), levels = 1:R)
#' print(sdcov(X, y))
#' print(sdcor(X, y))
#'
#' # Man-made functionally dependent data --------------------------------------
#' n <- 30; R <- 3; p <- 3
#' X <- matrix(0, n, p)
#' X[1:10, 1] <- 1; X[11:20, 2] <- 1; X[21:30, 3] <- 1
#' y <- factor(rep(1:3, each = 10))
#' print(sdcov(X, y))
#' print(sdcor(X, y))
#'
#' @export
sdcov <- function(X, y, type = "V", return_mat = FALSE) {
  if (is.data.frame(X)) {
    X <- as.matrix(X)
  } else if (is.numeric(X)) {
    X <- as.matrix(X, ncol = 1) # Univariate case
  } else if (!is.matrix(X)) {
    stop("`X` should be a matrix or a numeric vector")
  }
  
  if (!is.factor(y)) {
    stop("`y` should be a factor")
  } else {
    Y <- switch_cat_repr(y) # Convert to indicator matrix representation
  }
  
  n <- nrow(Y)
  n_r <- colSums(Y)
  
  if (type == "U") {
    Y <- Y[, which(n_r > 1), drop = FALSE] # Filter: `r` such that n_r > 1
    prob <- (colSums(Y) - 1) / (n - 1)
    
    denom <- n * (n - 1)
    mat_x <- as.matrix(dist(X))
    mat_y <- 1 - Y %*% diag(1 / prob) %*% t(Y)
  } else if (type == "V") {
    prob <- colSums(Y) / n
    
    denom <- n^2
    mat_x <- as.matrix(dist(X))
    # mat_x <- energy::Dcenter(X) # Equivalent!
    mat_y <- 1 - Y %*% diag(1 / prob) %*% t(Y)
  } else {
    stop('`type` should be either "U" or "V"')
  }
  
  sdcov <- sum(mat_x * mat_y) / denom
  
  if (return_mat) {
    return(
      list(
        sdcov = sdcov,
        mat_x = mat_x,
        mat_y = mat_y
      )
    )
  } else {
    return(sdcov)
  }
}

#' @rdname sdcov
#' @export
sdcor <- function(X, y) {
  R <- length(levels(y))
  
  dvar_X <- energy::dcov(X, X)
  sdcor <- sdcov(X, y) / (dvar_X * sqrt(R - 1))
  
  sdcor
}

#' Semi-distance independence test
#'
#' @description Implement the semi-distance independence test via permutation
#'   test, or via the asymptotic approximation when the dimensionality of
#'   continuous variables \eqn{p} is high.
#'
#' @inheritParams sdcov
#' @param test_type Type of the test:
#'   * `"perm"` (the default): Implement the test via permutation test;
#'   * `"asym"`: Implement the test via the asymptotic approximation when the
#'   dimension of continuous variables \eqn{p} is high.
#'
#'   See the Reference for details.
#' @param num_perm The number of replications in permutation test. Defaults to
#'   10000. See Details and Reference.
#'
#' @details The semi-distance independence test statistic is \deqn{T_n = n \cdot
#'   \widetilde{\text{SDcov}}_n(X, y),} where the
#'   \eqn{\widetilde{\text{SDcov}}_n(X, y)} can be computed by `sdcov(X, y, type
#'   = "U")`.
#'
#'   For the permutation test (`test_type = "perm"`), totally \eqn{K}
#'   replications of permutation will be conducted, and the argument `num_perm`
#'   specifies the \eqn{K} here. The p-value of permutation test is computed by
#'   \deqn{\text{p-value} = (\sum_{k=1}^K I(T^{\ast (k)}_{n} \ge T_{n}) + 1) /
#'   (K + 1),} where \eqn{T_{n}} is the semi-distance test statistic and
#'   \eqn{T^{\ast (k)}_{n}} is the test statistic with \eqn{k}-th permutation
#'   sample.
#'
#'   When the dimension of the continuous variables is high, the asymptotic
#'   approximation approach can be applied (`test_type = "asym"`), which is
#'   computationally faster since no permutation is needed.
#'
#' @returns A list with class `"indtest"` containing the following components
#' * `method`: name of the test;
#' * `name_data`: names of the `X` and `y`;
#' * `n`: sample size of the data;
#' * `test_type`: type of the test;
#' * `num_perm`: number of replications in permutation test, if
#'   `test_type = "perm"`;
#' * `stat`: test statistic;
#' * `pvalue`: computed p-value.
#'
#' @seealso [sdcov()] for computing the statistic of semi-distance covariance.
#'
#' @examples
#' X <- mtcars[, c("mpg", "disp", "drat", "wt")]
#' y <- factor(mtcars[, "am"])
#' test <- sd_test(X, y)
#' print(test)
#'
#' # Man-made independent data -------------------------------------------------
#' n <- 30; R <- 5; p <- 3; prob <- rep(1/R, R)
#' X <- matrix(rnorm(n*p), n, p)
#' y <- factor(sample(1:R, size = n, replace = TRUE, prob = prob), levels = 1:R)
#' test <- sd_test(X, y)
#' print(test)
#'
#' # Man-made functionally dependent data --------------------------------------
#' n <- 30; R <- 3; p <- 3
#' X <- matrix(0, n, p)
#' X[1:10, 1] <- 1; X[11:20, 2] <- 1; X[21:30, 3] <- 1
#' y <- factor(rep(1:3, each = 10))
#' test <- sd_test(X, y)
#' print(test)
#'
#' #' Man-made high-dimensionally independent data -----------------------------
#' n <- 30; R <- 3; p <- 100
#' X <- matrix(rnorm(n*p), n, p)
#' y <- factor(rep(1:3, each = 10))
#' test <- sd_test(X, y)
#' print(test)
#'
#' test <- sd_test(X, y, test_type = "asym")
#' print(test)
#'
#' # Man-made high-dimensionally dependent data --------------------------------
#' n <- 30; R <- 3; p <- 100
#' X <- matrix(0, n, p)
#' X[1:10, 1] <- 1; X[11:20, 2] <- 1; X[21:30, 3] <- 1
#' y <- factor(rep(1:3, each = 10))
#' test <- sd_test(X, y)
#' print(test)
#'
#' test <- sd_test(X, y, test_type = "asym")
#' print(test)
#'
#' @export
sd_test <- function(X, y, test_type = "perm", num_perm = 10000) {
  obj <- sdcov(X, y, type = "U", return_mat = TRUE)
  
  n <- length(y)
  name_data <- paste(
    deparse(substitute(X)), "and", deparse(substitute(y))
  )
  
  sdcov_n <- obj$sdcov
  mat_x <- obj$mat_x
  mat_y <- obj$mat_y
  
  stat <- n * sdcov_n
  
  if (test_type == "perm") {
    name_method <- paste0(
      "Semi-Distance Independence Test (Permutation Test with K = ",
      num_perm, ")"
    )
    
    num_rej <- 0
    for (perm in 1:num_perm) {
      ind_perm <- sample(1:n)
      mat_y_perm <- mat_y[ind_perm, ind_perm]
      sdcov_star <- sum(mat_x * mat_y_perm)/ (n * (n - 1))
      if (sdcov_star >= sdcov_n) {
        num_rej <- num_rej + 1
      }
    }
    pvalue <- (num_rej + 1) / (num_perm + 1)
  } else if (test_type == "asym") {
    name_method <- "Semi-distance Independence (Asymptotic) Test"
    
    tr_obj <- tr_estimate(X)
    tr_S <- tr_obj$tr_S
    tr_S2 <- tr_obj$tr_S2
    
    n_r <- colSums(switch_cat_repr(y))
    d <- tr_S2 / tr_S * (sum(n_r / (n_r - 1)) - n / (n - 1))
    pvalue <- pnorm(stat, sd = sqrt(d), lower.tail = FALSE) # one-sided
  } else {
    stop('`test_type` should be either "perm" or "asym"!')
  }
  
  indtest <- list(
    method = name_method,
    name_data = name_data,
    n = n,
    test_type = test_type,
    stat = n * sdcov_n,
    pvalue = pvalue
  )
  if (test_type == "perm") {
    indtest$num_perm = num_perm
  }
  class(indtest) <- "indtest"
  
  indtest
}

#' Feature screening via semi-distance correlation
#'
#' @description Implement the (grouped) feature screening for the classification
#'   problem via semi-distance correlation.
#'
#' @param X Data of multivariate covariates, which should be an
#'   \eqn{n}-by-\eqn{p} matrix.
#' @param y Data of categorical response, which should be a factor of length
#'   \eqn{n}.
#' @param group_info A list specifying the group information, with elements
#'   being sets of indicies of covariates in a same group. For example,
#'   `list(c(1, 2, 3), c(4, 5))` specifies that covariates 1, 2, 3 are in a
#'   group and covariates 4, 5 are in another group.
#'
#'   Defaults to `NULL`. If `NULL`, then it will be set as `list(1, 2, ..., p)`,
#'   that is, treat each single covariate as a group.
#'
#'   If `X` has colnames, then the colnames can be used to specified the
#'   `group_info`. For example, `list(c("a", "b"), c("c", "d"))`.
#'
#'   The names of the list can help recoginize the group. For example,
#'   `list(grp_ab = c("a", "b"), grp_cd = c("c", "d"))`. If names of the list
#'   are not specified, `c("Grp 1", "Grp 2", ..., "Grp J")` will be applied.
#'
#' @param d An integer specifying *at least* how many (single) features should
#'   be kept after screening. For example, if `group_info = list(c(1, 2), c(3,
#'   4))` and `d = 3`, then all features 1, 2, 3, 4 must be selected since it
#'   should guarantee at least 3 features are kept.
#'
#'   Defaults to `NULL`. If `NULL`, then it will be set as \eqn{[n / log(n)]},
#'   where \eqn{[x]} denotes the integer part of x.
#'
#' @param parallel A boolean indicating whether to calculate parallelly via
#'   `furrr::future_map`. Defaults to `FALSE`.
#'
#' @returns A list of the objects about the implemented feature screening:
#' * `group_info`: group information;
#' * `measurement`: sample semi-distance correlations calculated for the groups
#'   specified in `group_info`;
#' * `selected`: indicies/names of (single) covariates that are selected after
#'   feature screening;
#' * `ordering`: order of the calculated measurements of the groups specified in
#'   `group_info`. The first one is the largest, and the last is the smallest.
#'
#' @seealso [sdcor()] for calculating the sample semi-distance correlation.
#'
#' @examples
#' X <- mtcars[, c("mpg", "disp", "hp", "drat", "wt", "qsec")]
#' y <- factor(mtcars[, "am"])
#'
#' sd_sis(X, y, d = 4)
#'
#' # Suppose we have prior information for the group structure as
#' # ("mpg", "drat"), ("disp", "hp") and ("wt", "qsec")
#' group_info <- list(
#'   mpg_drat = c("mpg", "drat"),
#'   disp_hp = c("disp", "hp"),
#'   wt_qsec = c("wt", "qsec")
#' )
#' sd_sis(X, y, group_info, d = 4)
#'
#' @export
sd_sis <- function(X, y, group_info = NULL, d = NULL, parallel = FALSE) {
  n <- nrow(X)
  p <- ncol(X)
  
  if (is.null(d)) {
    d <- floor(n / log(n))
  }
  
  if (is.null(colnames(X))) {
    names_X <- 1:p
    colnames(X) <- names_X
  } else {
    names_X <- colnames(X)
  }
  
  if (is.null(group_info)) {
    group_info <- split(names_X, 1:p)
    names(group_info) <- paste("Grp", names_X)
  } else if (is.null(names(group_info))) {
    names(group_info) <- paste("Grp", 1:length(group_info))
  }
  
  J <- length(group_info)
  
  if (parallel) {
    omega <- furrr::future_map_dbl(group_info, ~ sdcor(X[, .x, drop = FALSE], y))
  } else {
    omega <- map_dbl(group_info, ~ sdcor(X[, .x], y))
  }
  
  ordering <- order(omega, decreasing = TRUE)
  ordering_names <- names(group_info)[ordering]
  
  if (length(group_info[[ordering[1]]]) >= d) {
    D_hat <- group_info[[ordering[1]]]
  } else {
    D_hat <- c()
    for (j in 1:J) {
      D_hat <- unique(append(D_hat, group_info[[ordering[j]]]))
      if (length(D_hat) >= d) {
        break
      } 
    }
  }
  
  list(
    group_info = group_info,
    measurement = omega,
    selected = D_hat,
    ordering = ordering_names
  )
}
